### PR TITLE
Fix legacy backend crash on assert 

### PR DESCRIPTION
**Pull request by rartemev**
Fix for the issue #12464

### CODE REVIEW

**Comment by rartemev** [comment]
@dotnet-bot test Windows_NT x86lb Debug Build and Test

File: src/jit/codegenlegacy.cpp [hunk 1]
```diff
@@ -11364,8 +11364,6 @@ void CodeGen::genCodeForTreeSmpOpAsg(GenTreePtr tree)
 
             if (VarSetOps::IsMember(compiler, genUpdateLiveSetForward(op2), varDsc->lvVarIndex))
             {
-                noway_assert(compiler->opts.compDbgCode);
```
Comment (CarolEidt) [inline src/jit/codegenlegacy.cpp]: Do you understand why this assert is hitting now, but was (presumably) not hitting before? Is it a new assert, or has something else changed?

Comment (CarolEidt) [inline src/jit/codegenlegacy.cpp]: Note that this case will arise if we are assigning to a lclVar that is currently live, which shouldn't happen except in the case where we are artificially keeping lclVars live everywhere, which doesn't seem to be the case here. Have you checked that the liveness is computed correctly in this case?

Comment (rartemev) [inline src/jit/codegenlegacy.cpp]: I have no idea whether it is new or not.
My thought was like why we are checking for generating 'debug-safe' code under condition that local var alive inside expression. It has no sense for me so I decided to delete it.

Comment (CarolEidt) [inline src/jit/codegenlegacy.cpp]: The assert is saying that, unless this is debug code, we should never encounter a definition of a lclVar that is currently live (i.e. the definition should have killed any uses that are upwards-exposed at this point). So, if this assert fires there is either some valid reason we are keeping it alive, OR there is a liveness bug.

**Comment by rartemev** [comment]
@dotnet/jit-contrib PTAL

**Comment by rartemev** [comment]
@dotnet/jit-contrib PTAL. This fix blocks #13073

File: src/jit/flowgraph.cpp [hunk 2]
```diff
@@ -9315,7 +9315,7 @@ VARSET_VALRET_TP Compiler::fgGetVarBits(GenTreePtr tree)
     {
         VarSetOps::AddElemD(this, varBits, varDsc->lvVarIndex);
     }
-    else if (varDsc->lvType == TYP_STRUCT && varDsc->lvPromoted)
+    else if (tree->gtType == TYP_STRUCT && varDsc->lvPromoted)
```
Comment (CarolEidt) [inline src/jit/flowgraph.cpp]: What is the case where this was doing the wrong thing? Was it a `GT_LCL_FLD` tree that should not be affecting all the fields?

**Comment by BruceForstall** [comment]
Please explain what the problem was, and why this change is correct. Show JitDumps if necessary.

**Comment by BruceForstall** [comment]
Also, please run asm diffs to ensure no unexpected changes occur.

**Comment by rartemev** [comment]
@BruceForstall @CarolEidt 
```
N023( 58, 30) [000010] --CXG--------             /--* CALL r2r_ind ref System.Collections.Immutable.ImmutableInterlocked.InterlockedCompareExchange $182
N004(  9,  7) [000019] ------------- arg1 on STK |  +--* OBJ(4)    struct <l:$240, c:$201>
N003(  3,  3) [000018] L------------             |  |  \--* ADDR      byref  $1c1
N002(  3,  2) [000002] -------N-----             |  |     \--* LCL_VAR   struct(P) V01 arg1
                                                 |  |     \--* ref V01.array(offs=0x00) -> V04               $200
N007(  9,  7) [000016] U------------ arg2 on STK |  +--* OBJ(4)    struct (last use) <l:$241, c:$203>
N006(  3,  3) [000015] L------------             |  |  \--* ADDR      byref  $1c2
N005(  3,  2) [000009] -------N-----  >>1>>      |  |     \--* LCL_VAR   struct(U) V03 loc0   << Type of this was INT, not struct what actually is
                                                 |  |     \--* ref V03.array(offs=0x00) -> V05 tmp1          ecx $202
N012( 15,  7) [000013] --CXG--------             |  |  /--* CALL help r2r_ind int HELPER.CORINFO_HELP_READYTORUN_GENERIC_HANDLE $380
N010(  1,  1) [000011] ------------- arg0 in ecx |  |  |  \--* LCL_VAR   int V02 TypeCtx u:2 edx(last use) $100
N014( 15,  7) [000065] -ACXG---R-L-- arg3 SETUP  |  +--* ASG       int    $380
N013(  1,  1) [000064] D------N-----             |  |  \--* LCL_VAR   int V06 tmp2 d:3 eax $281
N019(  1,  1) [000066] ------------- arg3 in edx |  +--* LCL_VAR   int V06 tmp2 u:3 eax(last use) $281
N020(  1,  1) [000001] ------------- arg0 in ecx |  \--* LCL_VAR   byref V00 arg0 u:2 esi(last use) $80
N025( 58, 30) [000027] -ACXG---R----             * ASG       ref    $182
N024(  1,  1) [000024] D------N-----  >>2>>      \--* REG_VAR   ref V05 tmp1          ecx REG ecx RV $182
```
Due to the wrong VarDsc type struct's field haven't been processed correct so we had wrong livesets. Hence the V05 became alive when it actually wasn't.

**Comment by CarolEidt** [comment]
I don't think that this is the correct fix. IIUC you are saying that the the `LclVarDsc` for V03 is of type INT. However, it's not clear how that is possible if it has a field (`V03.array` which has apparently been promoted to `V05`. How is that happening?

**Comment by rartemev** [comment]
I don't know too. I've seen it under debbuger. Since I am not familiar with legacy backend I might only make an assumption that `.array` actually means INT_PTR which for some reasons became INT, when actual type is struct anyway

**Comment by CarolEidt** [comment]
What the dump is telling me is that V03 is a struct local, with a single field, named `array`, which is of type `ref`. V05 is the local for that field. The reason I infer that it has a single field is that the `OBJ` node is of size 4. None of that explains why the `LclVarDsc` for V03 would be `TYP_INT`. Is V03 a user local or argument? Or is it a compiler-created temp? What is its type when it is created.

**Comment by rartemev** [comment]
```
        public static bool A<T>(ref ImmutableArray<T> location, ImmutableArray<T> value)
        {
            return InterlockedCompareExchange(ref location, value, default(ImmutableArray<T>)).IsDefault;
        }
```
This is the C# source. So I think V3 has to be compiler-created temp. ImmutableArray<T> is a struct which contains `.array` field.

**Comment by CarolEidt** [comment]
@rartemev - You should be able to debug this - e.g. set a breakpoint at the return from `Compiler::lvaGrabTemp` when `tempNum` is 3. At that point, you should be able to see who is creating V03, and what they are setting its type to. Since we know from the dump that it is being promoted, at some point it must have been `TYP_STRUCT`. You should then be able to figure out where it is being changed to `TYP_INT`. I think that is probably where the bug is.

File: src/jit/flowgraph.cpp [hunk 3]
```diff
@@ -9315,8 +9315,14 @@ VARSET_VALRET_TP Compiler::fgGetVarBits(GenTreePtr tree)
     {
         VarSetOps::AddElemD(this, varBits, varDsc->lvVarIndex);
     }
-    else if (varDsc->lvType == TYP_STRUCT && varDsc->lvPromoted)
+    // We have to check type of root tree, not Local Var descriptor because
+    // for legacy backend we promote TYP_STRUCT to INT_INT in case unsed non-argument local.
```
Comment (CarolEidt) [inline src/jit/flowgraph.cpp]: Should be "we promote TYP_STRUCT to TYP_INT if it is an unused or independently promoted non-argument struct local"

**Comment by rartemev** [comment]
@dotnet-bot help

**Comment by rartemev** [comment]
@dotnet-bot help

**Pull request merged**